feat(blog): metrics blog post (#167)
someengineering/resoto.com: blog/2022/05-30-resoto-metrics/index.md

​kind=aws_ec2_instance, id=i-076b9763c755a9b51, name=16ca4.prod1, instance_cores=16, age=3yr2mo, cloud=aws, account=eng-production, region=us-west-2
​kind=aws_ec2_instance, id=i-074fcfe526f95c9fd, name=16ca5.prod1, instance_cores=16, age=3yr2mo, cloud=aws, account=eng-production, region=us-west-2
​kind=aws_ec2_instance, id=i-04e09d3c714048c4d, name=16ca6.prod1, instance_cores=16, age=3yr2mo, cloud=aws, account=eng-production, region=us-west-2
​kind=aws_ec2_instance, id=i-0d2dfda13e02b2b20, name=16ca7.prod1, instance_cores=16, age=2yr9mo, cloud=aws, account=eng-production, region=us-west-2
​...
```

This would return a list of all the EC2 instances with more than 4 cores. That's useful if I want to do something with each individual instance, but sometimes I'm not interested in the details of individual resources. Sometimes I just want to know the sum of how many resources there are. Or, how many resources of a certain kind are running. Like the distribution of compute instances by instance type (e.g. how many m5.large, m5.2xlarge, etc.) or the current cost of compute plus storage grouped by team.

#### Aggregating

> This is where the before mentioned [aggregation](/blog/2022/03/03/aggregating-search-data) comes into play. It does just that; [aggregating and grouping the results of a search](/blog/2022/03/03/aggregating-search-data). This is what creates the samples of a metric.

Reviewer: ```suggestion
[Aggregating and grouping the results of a search](/blog/2022/03/03/aggregating-search-data) creates the samples of a metric.
```

```
> search aggregate(/ancestors.cloud.reported.name as cloud, /ancestors.account.reported.name as account, /ancestors.region.reported.name as region, instance_type as type, instance_status as status: sum(1) as instances_total): is(instance)
​
​group:
​  cloud: aws
​  account: eng-production
​  region: us-west-2
​  type: m5.xlarge
​  status: running
​instances_total: 13
​
​---
​group:
​  cloud: aws
​  account: eng-production
​  region: us-west-2
​  type: m5.4xlarge
​  status: stopped
​instances_total: 7
​...
```

> Now this is useful but what would be even more useful was if I could compare the current values to the ones from an hour ago, a day ago, a month ago, a year ago, etc. This is where time series come into play.

Reviewer: ```suggestion
This is useful, but the ability to compare current values to those from an hour, day, month, year, etc. ago would be even more useful. This brings us to the next concept, time series.
```

#### Time series

> A time series database like [Prometheus](https://prometheus.io/) does not store the details of an individual resource but instead stores the aggregated data over time. It then allows us to query that data and create charts to visualize the result. In the aggregated search above each of the results is what Prometheus calls a sample. A sample is a single value at a point in time in a time series. In our previous example the `cloud`, `account`, `region`, `type` and `status` in each group is what's called a [label](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) in Prometheus. Labels are `key: value` pairs that allow us to group samples.

Reviewer: ```suggestion
Time series databases such as [Prometheus](https://prometheus.io) do not store details of individual resources, but aggregated data over time—allowing us to query aggregate data and create charts to visualize the results.

In the aggregated search above, each result is what [Prometheus](https://prometheus.io) calls a [sample](https://prometheus.io/docs/concepts/data_model#samples). A [sample](https://prometheus.io/docs/concepts/data_model#samples) is a single value at a specific point in time.

Looking again at the same example, `cloud`, `account`, `region`, `type`, and `status` in each group are [labels](https://prometheus.io/docs/concepts/data_model#metric-names-and-labels). [Labels](https://prometheus.io/docs/concepts/data_model#metric-names-and-labels) are `key: value` pairs that allow us to group [samples](https://prometheus.io/docs/concepts/data_model#samples).
```

> Prometheus has some basic graphing capabilities but to build a dashboard we want to use a better suited tool like [Grafana](https://grafana.com/). It can visualize data from different sources in a variety of chart styles, like this stacked line chart.

Reviewer: ```suggestion
Prometheus has basic graphing capabilities, but [Grafana](https://grafana.com) allows you to build a dashboard visualizing data from different sources in a variety of chart styles, like this stacked line chart:
```

```

> config edit resoto.metrics

```

```yaml
resotometrics:
  metrics:
    instances_total:
      # Metric help text
      help: 'Number of Instances'
      # Aggregation search to run
      search: 'aggregate(/ancestors.cloud.reported.name as cloud, /ancestors.account.reported.name as account, /ancestors.region.reported.name as region, instance_type as type, instance_status as status: sum(1) as instances_total): is(instance)'
      # Type of metric (gauge or counter)
      type: 'gauge'
...
```

Here you can add your own metrics. As [previously explained](#aggregating), the `aggregate` expression in the `search` field is what creates the samples of a metric.

This config can be updated at runtime. Next time the `metrics` [workflow](/docs/concepts/automation/workflow) is run Resoto Metrics will generate the new metric and from then on provide it to Prometheus.

```
> workflow run metrics

> ```

Author: Why do you prefix all resh commands with bash syntax? This is wrong, it's not bash!

Good, the data points are connected and averaged over time. However the amount of labels is a bit overwhelming. Right now we are seeing one stacked chart per unique label combination. Let's try to reduce the amount of labels by summing them all up.

```

sum(avg_over_time(resoto_instances_total{status="running"}[1h]))

```

![Prometheus summed metrics](img/prometheus_summed_metrics.png)

Nice, now we see how the total number of compute instances has changed over the last two weeks. However we lost absolutely all labels. No more accounts, region and instance type information. To get some information back, let's group the summed up averages by account.

```

sum(avg_over_time(resoto_instances_total{status="running"}[1h])) by (account)

```

![Prometheus summed metrics by account](img/prometheus_summed_metrics_by_account.png)

Neat, we see how the number of compute instances has changed over time for each account.

Want to see how storage has changed over time? Just change `resoto_instances_total` to `resoto_volume_bytes`. Want to see $$$ spent per hour? `resoto_instances_hourly_cost_estimate` is the metric you are looking for.

> ## How Metrics Are Made

Author: I'm trying to tell the reader about the Flowers and the Bees that make a metrics. Like at first you search the data, then you aggregate the data, then you attach a timestamp to the aggregation result and store that newly created concoction in a time series database. Then you take a whole bunch of these time series and apply a function like sum of average over time on them, voila a new metric is born.

Now, the Prometheus web UI will provide syntax help and auto-complete for available metric names. However, you might be wondering—how are you supposed to know which metrics exist? How do you know what other metrics there are and where something like `resoto_instances_total` is defined?

> Metrics are configured in the `resoto.metrics` [configuration](/docs/getting-started/configuring-resoto). Within [Resoto Shell (`resh`)](/docs/concepts/components/shell), simply execute:

Reviewer: ```suggestion
Metrics are defined in the `resoto.metrics` [configuration](/docs/getting-started/configuring-resoto). To edit metrics definitions, execute the following command in [Resoto Shell](/docs/concepts/components/shell):
```

```bash
> config edit resoto.metrics
```

```yaml
resotometrics:
  metrics:
    instances_total:
      # Metric help text
      help: 'Number of Instances'
      # Aggregation search to run
      search: 'aggregate(/ancestors.cloud.reported.name as cloud, /ancestors.account.reported.name as account, /ancestors.region.reported.name as region, instance_type as type, instance_status as status: sum(1) as instances_total): is(instance)'
      # Type of metric (gauge or counter)
      type: 'gauge'
...
```

> Here you can add your own metrics. As [previously explained](#aggregating), the `aggregate` expression in the `search` field is what creates the samples of a metric.

Reviewer: ```suggestion
As [described above](#aggregating), the `aggregate` expression in the `search` field is what creates the samples of a metric.
```

This config can be updated at runtime. Next time the `metrics` [workflow](/docs/concepts/automation/workflow) is run Resoto Metrics will generate the new metric and from then on provide it to Prometheus.

```bash
> workflow run metrics
```

## Creating the Metrics Dashboard

> So now we have learned how to get metrics from Resoto into Prometheus, how to query them and how to create new ones. But what about the dashboard?
> 
> Right, fasten your seatbelt. This will go fast.

Reviewer: ```suggestion
Now that we've learned how to get metrics from Resoto into [Prometheus](https://prometheus.io), query metrics, and define new metrics, we can create the dashboard.

Alright, fasten your seatbelts! This will go fast. 🏎️💨
```

> 1. Run the Grafana Docker container

Reviewer: ```suggestion
1. Start the [Grafana Docker container](https://hub.docker.com/r/grafana/grafana-oss):
```

   ```bash
   $ docker run -d -p 3000:3000 -v grafana-data:/var/lib/grafana -v grafana-etc:/etc/grafana grafana/grafana-oss
   ```

> 2. Open up the Grafana UI (e.g. [http://localhost:3000](http://localhost:3000))

Reviewer: ```suggestion
2. Open the Grafana UI (e.g., [http://localhost:3000](http://localhost:3000)).
```

> 3. Login as `admin` with password `admin` and set a new password.

Author: Idk. You don't really have a choice, it will ask you to change the password and you can't do anything else.

4. On the left go to Settings > Data Sources > Add Data Source > Prometheus

> 5. In the URL field, enter the Prometheus URL e.g. `http://tsdb.docker.internal:9090`

Author: Do we really want to link to Prometheus 100 times in this post?

   ![Add Prometheus Data Source to Grafana](img/grafana_setup1.png)

> 6. Scroll down and click `Save & test`. Make sure it replies with "Data source is working".

Reviewer: ```suggestion
6. Scroll down and click the **Save & test** button. Make sure that the result is "Data source is working":
```

   ![Save Data Source](img/grafana_setup2.png)

7. On the left click on the `+` button and select `Create Dashboard`, then click the `Save` button in the top menu bar.

   ![Save dashboard](img/grafana_save_dashboard.png)

> 8. On the top click on `Dashboard settings` > `Variables` and `Add variable`.

Reviewer: ```suggestion
8. Select **Dashboard settings > Variables** and click the **Add variable** button:
```
